### PR TITLE
Add binlog filepath uniquification option

### LINKED ISSUE #8817: [Feature Request]: Opt-in mechanism for unique default names of binlogs
### Summary

If explicit name for binlog is not specified `msbuild.binlog` is being used - which can lead to silent overwrites of pre-existing binlogs. This is especially painfull in more complicated workflows where msbuild is being invoked multipe times.

The cases where we heared about this could be useful:
* Failing to build C# project with fakes, due to attempts for concurrent writes to msbuild.binlog (https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1791139) FYI @ocallesp 
* Onboard binlogs in AzDO FYI @Json

### Proposed Feature

Add opt-in mechanism (probably commandline switch + environment variable), which would turn on generation of unique default names for binlogs

### Alternative Designs

_No response_

### PR DESCRIPTION
**Pull request by JanKrivanek**
Fixes #8817

### Context
There are cases where explicit unique binlog paths cannot be specified and the implicit overwriting behavior is complicating the invectigations (e.g. the implicit separate restore step invoked by sdk - as per #9614).

For those cases let's introduce a simplistic uniquification opt-in mechanism: When specifying a '{}' substring withing the binlog path parameter - this is going to be expanded into a timestamp plus random stamp.

Behavior:

* Only '{}' is recognized and expanded. If the brackets have any content - it's intentionally left untouched (this can be added in future iteration if there is any need)
* If '{}' is present, then the '.binlog' extension is ensured (whether specified in original name or not)
* Each '{}' in the binlog path argument is replaced with timestamp and random string stamp
* Custom expander can be injected - this is mainly for unit tests, but can be used if anyone wants different uniquification
* Intentionally no build-time information (entrypoint project, target, etc) is added - KISS principle. All needed information can be located within the binlog. We can iterate on this if requested and needed.

Samples:

* `-bl:{}` --> `20240423-102016--16636--anxG+w.binlog`
* `-bl:LogFile=MyBinlog-{}.binlog` --> `MyBinlog-20240423-102016--16636--anxG+w.binlog`
* `-bl:{something}.binlog` --> `{something}.binlog`

### Changes Made
The parameters parsing is extended, so that it can recognize and expand '{}'

### Testing
Existing tests. + added tailored tests for pre-existing and new naming interpretation

### Notes
This can technically be breaking - but the likelyhood is quite low (I tried to search GH and ADO for usage of binlog, where explicit path would be specified and contained '{}' - and found none). S I'd prefer clean experience rather then adding additional commandline arguments. It is still behind the change wave - if anything goes wrong.

FYI @KirillOsenkov

### CODE REVIEW

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 1]
```diff
@@ -451,5 +451,69 @@ private void ProcessParameters(out bool omitInitialInfo)
...
+            {
+                return hasProperExtension;
+            }
+
+            filePath = parameter.Replace("{}", GetUniqueStamp(), StringComparison.Ordinal);
+
+            if (!hasProperExtension)
+            {
+                filePath += ".binlog";
+            }
```
Comment (f-alizada) [inline src/Build/Logging/BinaryLogger/BinaryLogger.cs:480]: If the parameter presented but with incorrect extension, do we want to fail fast and report the error?

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 2]
```diff
@@ -451,5 +451,69 @@ private void ProcessParameters(out bool omitInitialInfo)
...
+
+        private string GetUniqueStamp()
+            => PathParameterExpander.ExpandParameter(string.Empty);
+
+        private class BinlogPathParameterExpander : IBinlogPathParameterExpander
+        {
+            public string ExpandParameter(string parameters)
+                => $"{DateTime.UtcNow.ToString("yyyyMMddHHmmss")}-{GenerateRandomString(6)}";
+
+            private string GenerateRandomString(int length)
```
Comment (f-alizada) [inline src/Build/Logging/BinaryLogger/BinaryLogger.cs:492]: Shall we move this method to shared codebase helpers utils or something similar to have a possibility to reuse it by contributors? :)

File: src/Build.UnitTests/BinaryLogger_Tests.cs [hunk 3]
```diff
@@ -637,6 +639,61 @@ public void SuppressCommandOutputForNonDiagVerbosity()
...
+            var eventSource = A.Fake<IEventSource>();
+
+            binaryLogger.Initialize(eventSource);
+            string expectedLog = Path.GetFullPath(expectedBinlogFile);
+            binaryLogger.FilePath.Should().BeEquivalentTo(expectedLog);
+            binaryLogger.Shutdown();
+            File.Exists(binaryLogger.FilePath).ShouldBeTrue();
+            FileUtilities.DeleteNoThrow(binaryLogger.FilePath);
+
+            File.Create(_logFile);
```
Comment (f-alizada) [inline src/Build.UnitTests/BinaryLogger_Tests.cs:674]: Could you please clarify why we are creating _logFile here?

**Review by f-alizada — APPROVED** [review]
Sorry for delaying with the PR and proper fix for the opt-in mechanism, the changes looks good. 
If the completely opt-in mechanism will be required please let me know, I have some draft changes that could be used as well :)

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 4]
```diff
@@ -135,6 +135,12 @@ public enum ProjectImportsCollectionMode
         /// </summary>
         public string Parameters { get; set; }
 
+        /// <summary>
+        /// Optional expander of wildcard(s) within the path parameter of a binlog <see cref="Parameters"/>.
+        /// See <see cref="IBinlogPathParameterExpander"/> for more details.
+        /// </summary>
+        public IBinlogPathParameterExpander PathParameterExpander { private get; set; } = new BinlogPathParameterExpander();
```
Comment (KirillOsenkov) [inline src/Build/Logging/BinaryLogger/BinaryLogger.cs:142]: I thought you said simplistic ;) Just use a `Func<string>` here (or `Func<string, string>`) and keep the default value null.

If there are currently no callers setting it from outside, we should make it internal.

**Comment by KirillOsenkov** [comment]
I like the design!

I feel pretty strongly that we should remove the interface and just use `Func<string>` to obtain the random string. If it's null, fall back to the default (current) implementation. 

Nothing in this PR should be adding any new public API.

Also would be nice if the process ID of the node was featured in the random string.

Also let's add a dash between date and time.

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs (hunk 4 above)
Comment (JanKrivanek) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:142]: Absolutely.. it was just my excessive love for interfaces. And then I needed to expose it due to easier mocking.
Lm flip to func - which will solve both of those

File: src/Build.UnitTests/BinaryLogger_Tests.cs (hunk 3 above)
Comment (JanKrivanek) [reply src/Build.UnitTests/BinaryLogger_Tests.cs:674]: Added the clarifying comment

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs (hunk 2 above)
Comment (JanKrivanek) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:492]: I moved it to Build.Framework

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs (hunk 1 above)
Comment (JanKrivanek) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:480]: In this case I'd preffer more friendly behavior - if the user forget to add the extension, it's automatically added.

I added it only for the case of wildcarded paths - not to break the general existing path... but I believe that in the short future we should have as general behavior (there is likely no good in throwing exception telling you you should add an extension)

**Comment by JanKrivanek** [comment]
> I like the design!
> 
> I feel pretty strongly that we should remove the interface and just use `Func<string>` to obtain the random string. If it's null, fall back to the default (current) implementation.
> 
> Nothing in this PR should be adding any new public API.
> 
> Also would be nice if the process ID of the node was featured in the random string.
> 
> Also let's add a dash between date and time.

Update format based on suggestions - the sample result is in the updated PR description

File: src/Build.UnitTests/BinaryLogger_Tests.cs [hunk 5]
```diff
@@ -637,6 +639,61 @@ public void SuppressCommandOutputForNonDiagVerbosity()
...
+            binaryLogger.PathParameterExpander = _ => random;
+
+            var eventSource = A.Fake<IEventSource>();
+
+            binaryLogger.Initialize(eventSource);
+            string expectedLog = Path.GetFullPath(expectedBinlogFile);
+            binaryLogger.FilePath.Should().BeEquivalentTo(expectedLog);
+            binaryLogger.Shutdown();
+            File.Exists(binaryLogger.FilePath).ShouldBeTrue();
+            FileUtilities.DeleteNoThrow(binaryLogger.FilePath);
```
Comment (surayya-MS) [inline src/Build.UnitTests/BinaryLogger_Tests.cs:670]: This was confusing at first - checking the file is there, then deleting it and creating again for the invariant in the constructor. Other tests use env to create `TransientTestFolder`. 
As a suggestion: also set the directory of the binlog file as a `TransientTestFolder`

**Review by surayya-MS — APPROVED** [review]

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 6]
```diff
@@ -451,5 +454,47 @@ private void ProcessParameters(out bool omitInitialInfo)
...
+                filePath += ".binlog";
+            }
+            return true;
+        }
+
+        private string GetUniqueStamp()
+            => (PathParameterExpander ?? ExpandPathParameter)(string.Empty);
+
+        private static string ExpandPathParameter(string parameters)
+            => $"{DateTime.UtcNow.ToString("yyyyMMdd-HHmmss")}--{ProcessId}--{StringUtils.GenerateRandomString(6)}";
```
Comment (surayya-MS) [inline src/Build/Logging/BinaryLogger/BinaryLogger.cs:491]: Also, out of curiosity, why did you decide to use the pattern of date/time/process id/string? Why not guid for example?

Comment (JanKrivanek) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:491]: Guid would do the trick as well.
The time and process-id is bit more descriptive and together with the randomized string brings same disambiguation result

Comment (baronfel) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:491]: I strongly want to push back on GUIDs - they are actively user hostile and don't help the user identity which binlog matches with which build invocation.

**Review by KirillOsenkov — APPROVED** [review]

File: src/Build.UnitTests/BinaryLogger_Tests.cs (hunk 5 above)
Comment (JanKrivanek) [reply src/Build.UnitTests/BinaryLogger_Tests.cs:670]: Unfortunately the TransientTestFolder cannot be used here.
I want to test here absolute and as well relative paths - with and without nesting. Relative paths are 'absolutized' by the logger - relative to the current execution folder. That means that the resulting log can be present within the bin folder - one that we should not (and won't be able to) delete

**Comment by rainersigwald** [comment]
Finally had a need to use this and unfortunately powershell eats the `{}` so you have to call, for example, `msbuild -m '-bl:{}'`.

**Comment by baronfel** [comment]
*shakes fist* POWERSHELLLLL

**Comment by JanKrivanek** [comment]
> Finally had a need to use this and unfortunately powershell eats the `{}` so you have to call, for example, `msbuild -m '-bl:{}'`.

Or escape `msbuild -m -bl:{{}}`

**Pull request merged**